plugins/common/starlark/starlark.go

```diff
@@ -0,0 +1,154 @@
+	if s.Source == "" && s.Script == "" {
+		return nil, errors.New("one of source or script must be set")
+	}
+	if s.Source != "" && s.Script != "" {
+		return nil, errors.New("both source or script cannot be set")
+	}
+
+	s.thread = &starlark.Thread{
+		Print: func(_ *starlark.Thread, msg string) { s.Log.Debug(msg) },
+		Load: func(thread *starlark.Thread, module string) (starlark.StringDict, error) {
+			return s.StarlarkLoadFunc(module, s.Log)
+		},
+	}
+
+	builtins := starlark.StringDict{}
+	builtins["Metric"] = starlark.NewBuiltin("Metric", newMetric)
+	builtins["deepcopy"] = starlark.NewBuiltin("deepcopy", deepcopy)
+	builtins["catch"] = starlark.NewBuiltin("catch", catch)
+	s.addConstants(&builtins)
+
+	program, err := s.sourceProgram(builtins, "")
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Execute source
+	globals, err := program.Init(s.thread, builtins)
+	if err != nil {
+		return nil, err
+	}
+	// Make available a shared state to the apply function
+	globals["state"] = starlark.NewDict(0)
+
+	// Freeze the global state.  This prevents modifications to the processor
+	// state and prevents scripts from containing errors storing tracking
+	// metrics.  Tasks that require global state will not be possible due to
+	// this, so maybe we should relax this in the future.
+	globals.Freeze()
+
+	return globals, nil
+}
+
+func InitFunction(globals starlark.StringDict, fnName string, expectedParams int) (*starlark.Function, error) {
+	globalFn := globals[fnName]
+
+	if globalFn == nil {
+		return nil, fmt.Errorf("%s is not defined", fnName)
+	}
+
+	var ok bool
+	var fn *starlark.Function
+	if fn, ok = globalFn.(*starlark.Function); !ok {
+		return nil, fmt.Errorf("%s is not a function", fnName)
+	}
+
+	if fn.NumParams() != expectedParams {
+		return nil, fmt.Errorf("%s function must take %d parameter(s)", fnName, expectedParams)
+	}
+	return fn, nil
+}
+
+// Add all the constants defined in the plugin as constants of the script
+func (s *StarlarkCommon) addConstants(builtins *starlark.StringDict) {
+	for key, val := range s.Constants {
+		sVal, err := asStarlarkValue(val)
+		if err != nil {
+			s.Log.Errorf("Unsupported type: %T", val)
+		}
+		(*builtins)[key] = sVal
+	}
+}
+
+func (s *StarlarkCommon) sourceProgram(builtins starlark.StringDict, filename string) (*starlark.Program, error) {
+	if s.Source != "" {
+		_, program, err := starlark.SourceProgram(filename, s.Source, builtins.Has)
+		return program, err
+	}
+	_, program, err := starlark.SourceProgram(s.Script, nil, builtins.Has)
+	return program, err
+}
+
+// Call calls the function fn with the specified positional and keyword arguments.
+func (s *StarlarkCommon) Call(fn starlark.Value, args starlark.Tuple) (starlark.Value, error) {
+	rv, err := starlark.Call(s.thread, fn, args, nil)
+	if err != nil {
+		if err, ok := err.(*starlark.EvalError); ok {
+			for _, line := range strings.Split(err.Backtrace(), "\n") {
+				s.Log.Error(line)
+			}
+		}
+	}
+	return rv, err
+}
```

Comment:
If you change the internal function and parameter representation to a map you could do
```suggestion
func (s *StarlarkCommon) Call(name string) (starlark.Value, error) {
	fn, ok := s.functions[name]
	if !ok {
		return nil, fmt.Errorf("function %q does not exist", name)
	}
	args, ok := s.parameters[name]
	if !ok {
		return nil, fmt.Errorf("params for function %q do not exist", name)
	}
	
	rv, err := starlark.Call(s.thread, fn, args, nil)
	if err != nil {
		if err, ok := err.(*starlark.EvalError); ok {
			for _, line := range strings.Split(err.Backtrace(), "\n") {
				s.Log.Error(line)
			}
		}
	}
	return rv, err
}
```